### PR TITLE
refactor in aux::write_ to avoid explicit castings

### PR DESCRIPTION
**Pull request by aldenml**

### CODE REVIEW

File: include/libtorrent/aux_/io.hpp [hunk 1]
```diff
@@ -59,12 +59,16 @@ namespace libtorrent { namespace aux
 		return ret;
 	}
 
-	template <class T, class Byte>
+	template <class T, class In, class Byte, typename Cond
+		= typename std::enable_if<std::is_integral<In>::value
```
Comment (arvidn) [inline include/libtorrent/aux_/io.hpp]: and ``T`` should probably be integral too

Comment (aldenml) [inline include/libtorrent/aux_/io.hpp]: `sizeof(In) == 1`? The type `In` is the one we don't want to cast, the one that is restricted already in the `enable_if`.

I will add the condition to `T`.

File: include/libtorrent/aux_/io.hpp [hunk 2]
```diff
@@ -59,12 +59,16 @@ namespace libtorrent { namespace aux
 		return ret;
 	}
 
-	template <class T, class Byte>
+	template <class T, class In, class Byte, typename Cond
+		= typename std::enable_if<std::is_integral<T>::value &&
+		(std::is_integral<In>::value || std::is_enum<In>::value)>::type>
 	inline typename std::enable_if<sizeof(Byte)==1, void>::type
-	write_impl(T val, span<Byte>& view)
+	write_impl(In data, span<Byte>& view)
```
Comment (arvidn) [inline include/libtorrent/aux_/io.hpp]: I meant to say ``sizeof(Byte)`` must be 1. The code in here assumes that it has 8 bits specifically.

**Comment by arvidn** [comment]
I guess ``Byte`` wasn't introduced in this patch. but would you mind adding another constraint?

File: include/libtorrent/aux_/io.hpp (hunk 2 above)
Comment (aldenml) [inline include/libtorrent/aux_/io.hpp]: I think that constraint is already there with the return type

**Comment by aldenml** [comment]
I don't mind at all, but see my comment above , i don't want to introduce redundant conditions

File: include/libtorrent/aux_/io.hpp (hunk 2 above)
Comment (arvidn) [inline include/libtorrent/aux_/io.hpp]: ah, I see

**Comment by arvidn** [comment]
lgtm

**Pull request merged**